src/viewport/services/auth_rate_limiter.py

```
@@ -0,0 +1,429 @@
+            raise InvalidClientAddressError("Malformed X-Forwarded-For header")
+        hops.append(_canonical_ip(element))
+    return tuple(hops)
+
+
+def _header_chain(request: Request) -> tuple[IPAddress, ...]:
+    forwarded_values = request.headers.getlist("forwarded")
+    xff_values = request.headers.getlist("x-forwarded-for")
+    forwarded = _parse_forwarded_header(",".join(forwarded_values)) if forwarded_values else None
+    xff = _parse_x_forwarded_for(",".join(xff_values)) if xff_values else None
+    if forwarded is not None and xff is not None and forwarded != xff:
+        raise InvalidClientAddressError("Forwarded headers disagree")
```

Ruff (0.16.1)

[warning] Avoid specifying long messages outside the exception class

(TRY003)